### PR TITLE
Allow constant reloading to load modules transitively.

### PR DESCRIPTION
**Pull request by mijoharas**
### Summary

Allow modules to be loaded transitively. Fixes https://github.com/rails/rails/issues/28997.

### CODE REVIEW

**Comment by pixeltrix** [comment]
This seems like a lot more work? Have you tried benchmarking it?

**Comment by mijoharas** [comment]
I haven't. How would you suggest going about that?

**Comment by mijoharas** [comment]
(also, I'd argue that you're only doing a lot more work immediately before you throw a constant missing error. So unless people are relying on those errors and catching them, they shouldn't see any slowdown).

**Comment by mijoharas** [comment]
@pixeltrix So I had a little glance around, reading the [contributers](http://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html) guide which suggests [this benchmarking script](https://github.com/rails/rails/blob/master/guides/bug_report_templates/benchmark.rb).

The section says:

>  The template is designed for testing relatively self-contained changes that can be inlined into the script.

Is that the template I should be using to benchmark the change? I'm not sure how I should benchmark it could you give me some pointers?

**Comment by pixeltrix** [comment]
@mijoharas yes, use that script as a starting point.

**Comment by mijoharas** [comment]
I've written a benchmark, the difference within the timings falls within error (luckily it allowed me to spot a small bug which I fixed):

```
Warming up --------------------------------------
            original   341.000  i/100ms
             updated   363.000  i/100ms
Calculating -------------------------------------
            original      3.377k (± 3.9%) i/s -     17.050k in   5.056942s
             updated      3.633k (± 4.7%) i/s -     18.513k in   5.106744s

Comparison:
             updated:     3633.1 i/s
            original:     3377.0 i/s - same-ish: difference falls within error
```

The benchmark that I ran is this:

```
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", github: "rails/rails"
  # gem "rails", path: '/home/michael/scripts/rails'
  gem "benchmark-ips"
  gem 'pry'
  gem 'pry-byebug'
end

require "active_support"
require "active_support/core_ext/object/blank"


# require 'active_support/dependencies.rb'
module ActiveSupport
  module Dependencies
    # Load the constant named +const_name+ which is missing from +from_mod+. If
    # it is not possible to load the constant into from_mod, try its parent
    # module using +const_missing+.
    def load_missing_constant_new(from_mod, const_name)
      unless qualified_const_defined?(from_mod.name) && Inflector.constantize(from_mod.name).equal?(from_mod)
        raise ArgumentError, "A copy of #{from_mod} has been removed from the module tree but is still active!"
      end

      # first check this module
      qualified_name = qualified_name_for from_mod, const_name
      visited_modules = Set.new
      if mod = load_missing_constant_from_file(from_mod, qualified_name, const_name)
        return mod
      # next check our parents
      elsif mod = load_missing_constant_from_parents(from_mod.parent, visited_modules, const_name)
        return mod
      # and then check other modules we have included
      elsif mod = load_missing_constant_from_ancestors(from_mod, visited_modules, const_name)
        return mod
      end

      name_error = NameError.new("uninitialized constant #{qualified_name}", const_name)
      name_error.set_backtrace(caller.reject { |l| l.starts_with? __FILE__ })
      raise name_error
    end

    def load_missing_constant_from_parents(from_mod, visited_modules, const_name)
      qualified_name = qualified_name_for from_mod, const_name
      if mod = load_missing_constant_from_file(from_mod, qualified_name, const_name)
        return mod
      elsif (parent = from_mod.parent) && parent != from_mod &&
            ! from_mod.parents.any? { |p| p.const_defined?(const_name, false) }
        # If our parents do not have a constant named +const_name+ then we are free
        # to attempt to load upwards. If they do have such a constant, then this
        # const_missing must be due to from_mod::const_name, which should not
        # return constants from from_mod's parents.
        begin
          # Since Ruby does not pass the nesting at the point the unknown
          # constant triggered the callback we cannot fully emulate constant
          # name lookup and need to make a trade-off: we are going to assume
          # that the nesting in the body of Foo::Bar is [Foo::Bar, Foo] even
          # though it might not be. Counterexamples are
          #
          #   class Foo::Bar
          #     Module.nesting # => [Foo::Bar]
          #   end
          #
          # or
          #
          #   module M::N
          #     module S::T
          #       Module.nesting # => [S::T, M::N]
          #     end
          #   end
          #
          # for example.
          visited_modules << parent
          return load_missing_constant_from_parents(parent, visited_modules, const_name)
        rescue NameError => e
          raise unless e.missing_name? qualified_name_for(parent, const_name)
        end
      end
    end

    def load_missing_constant_from_ancestors(from_mod, visited_modules, const_name)
      begin
        qualified_name = qualified_name_for from_mod, const_name
      rescue ArgumentError
        return nil
      end
      if mod = load_missing_constant_from_file(from_mod, qualified_name, const_name)
        return mod
      else
        visited_modules << from_mod
        not_yet_visited = from_mod.ancestors.reject do |x|
          visited_modules.include?(x)
        end
        mapped_search = not_yet_visited.lazy.map do |x|
          load_missing_constant_from_ancestors(x, visited_modules, const_name)
        end
        mapped_search.drop_while do |x|
          !x
        end.first
      end
    end

    def load_missing_constant_from_file(from_mod, qualified_name, const_name)
      path_suffix = qualified_name.underscore

      file_path = search_for_file(path_suffix)

      if file_path
        expanded = File.expand_path(file_path)
        expanded.sub!(/\.rb\z/, "".freeze)

        if loading.include?(expanded)
          raise "Circular dependency detected while autoloading constant #{qualified_name}"
        else
          require_or_load(expanded, qualified_name)
          raise LoadError, "Unable to autoload constant #{qualified_name}, expected #{file_path} to define it" unless from_mod.const_defined?(const_name, false)
          return from_mod.const_get(const_name)
        end
      elsif mod = autoload_module!(from_mod, const_name, qualified_name, path_suffix)
        return mod
      end
    end
  end
end



ActiveSupport::Dependencies.autoload_paths << '/tmp'



# search for needle located in module A
module A
  module B
    module C
      module D
        module E
          def self.test_one
            ActiveSupport::Dependencies.load_missing_constant(self, 'Needle')
            A.send(:remove_const, :Needle)
            ActiveSupport::Dependencies.loaded = Set.new
          end
          def self.test_two
            ActiveSupport::Dependencies.load_missing_constant_new(self, 'Needle')
            A.send(:remove_const, :Needle)
            ActiveSupport::Dependencies.loaded = Set.new
          end
        end
      end
    end
  end
end

Benchmark.ips do |x|
  x.report('original') { A::B::C::D::E.test_one }
  x.report('updated') { A::B::C::D::E.test_two }
  x.compare!
end
```

This benchmark requires one additional file to run, which is (`/tmp/a/needle.rb`):
```
module A
  class Needle
  end
end
```

**Comment by mijoharas** [comment]
To explain the benchmark file:
* First, we monkey patch our new methods into `Dependency`.
* We then add `/tmp` to our list of `autoload_paths` so that we can find the file `/tmp/a/needle.rb` (which has the contents mentioned above, to define the constant we are looking for `Needle` inside the module `A`).
* We then create a nested module structure, so that we can exercise our function a little bit and define our method to search for the constant within the nested module structure (making sure to clear our loaded list and constant so that we can try again).
* Then we run the benchmark calling the functions defined in the nested module structure over and over.

**Comment by mijoharas** [comment]
@pixeltrix let me know if we need anything further. I tried to write the benchmark to exercise the main part of the algorithm (the recursion to the parent module scopes). It doesn't exercise the work in searching the ancestor chain, as that is new functionality, so we can't compare it to the old version.

**Comment by mijoharas** [comment]
Hi @pixeltrix. Was wondering if there was anything else needed for this. (Sorry about the giant comment with lots of code.)

**Comment by mijoharas** [comment]
TL;DR wrote benchmark, it is as fast as the old one. In addition, I think the code is simpler now (obviously, that's opinion based). 

Does this PR need anything else?

**Comment by mijoharas** [comment]
Closing as this will be fixed with the new autoloading system
